### PR TITLE
Use ValueTask<T> on allocation hot paths

### PR DESCRIPTION
**Pull request by sharwell**
AnalyzerRunner indicated that these asynchronous paths typically complete synchronously. Measured savings are ~2.5% (~2GiB).

<details><summary>Ask Mode template not completed</summary>

### Customer scenario

What does the customer do to get into this situation, and why do we think this
is common enough to address for this release.  (Granted, sometimes this will be
obvious "Open project, VS crashes" but in general, I need to understand how
common a scenario is)

### Bugs this fixes

(either VSO or GitHub links)

### Workarounds, if any

Also, why we think they are insufficient for RC vs. RC2, RC3, or RTW

### Risk

This is generally a measure our how central the affected code is to adjacent
scenarios and thus how likely your fix is to destabilize a broader area of code

### Performance impact

(with a brief justification for that assessment (e.g. "Low perf impact because no extra allocations/no complexity changes" vs. "Low")

### Root cause analysis

How did we miss it?  What tests are we adding to guard against it in the future?

### How was the bug found?

(E.g. customer reported it vs. ad hoc testing)

### Test documentation updated?

If this is a new non-compiler feature or a significant improvement to an existing feature, update https://github.com/dotnet/roslyn/wiki/Manual-Testing once you know which release it is targeting.

</details>

### CODE REVIEW

File: src/Compilers/Core/Portable/CodeAnalysis.csproj [hunk 1]
```diff
@@ -63,6 +63,7 @@
     <PackageReference Include="System.ValueTuple" Version="$(SystemValueTupleVersion)" />
     <PackageReference Include="System.Xml.XPath.XDocument" Version="$(SystemXmlXPathXDocumentVersion)" />
     <PackageReference Include="System.Text.Encoding.CodePages" Version="$(SystemTextEncodingCodePagesVersion)" />
+    <PackageReference Include="System.Threading.Tasks.Extensions" Version="$(SystemThreadingTasksExtensionsVersion)" />
```
Comment (jasonmalinowski) [inline src/Compilers/Core/Portable/CodeAnalysis.csproj]: Does this need updates to the VisualStudioSetup.Dependencies package? Will it ship a new façade that we need to be ready for in RPS?

**Comment by sharwell** [comment]
@jasonmalinowski I meant to mark this one as a work in progress, to be revisited after merging the other allocation-reducing changes that don't require changes to dependencies.

File: src/Compilers/Core/Portable/CodeAnalysis.csproj (hunk 1 above)
Comment (jasonmalinowski) [reply src/Compilers/Core/Portable/CodeAnalysis.csproj]: (VS integration test failures imply yes, this needs more façade updates.)

**Comment by jasonmalinowski** [comment]
@sharwell Makes sense. No problem with this going in assuming the paperwork is done though. It'll be good to start getting ValueTask in here.

**Comment by CyrusNajmabadi** [comment]
How are you figuring out which paths to ValueTaskisfy?  It seems like thsi would be useful across a lot of the LS.  But it's unclear how to drive that, and my concern would be regressing some path that actually needs real async.

**Comment by sharwell** [comment]
> How are you figuring out which paths to ValueTaskisfy? 

In this case, I'm using AnalyzerRunner and looking for `Task<T>` allocations where the async state machine method `SetResult` is called by `Start`.

**Comment by sharwell** [comment]
@jasonmalinowski My original evaluation of allocation savings understated the value of this pull request by a factor of at least 4.

**Comment by jaredpar** [comment]
We need to make sure that we discuss the RPS implications of this with @MattGertz. This is a new library reference and very likely to be loading in our measured scenarios.

**Comment by sharwell** [comment]
@jaredpar Yes. I plan to update this pull request as issues come to light until it gets to a point that we can claim a justifiable win. However, I'm having all sorts of problems just getting the tests to pass with it.

File: src/Compilers/Core/Portable/CodeAnalysis.csproj (hunk 1 above)
Comment (tmat) [reply src/Compilers/Core/Portable/CodeAnalysis.csproj]: Likely.

Comment (tmat) [reply src/Compilers/Core/Portable/CodeAnalysis.csproj]: ... and binding redirects.

**Comment by sharwell** [comment]
@dotnet-bot retest windows_debug_unit32_prtest please

**Comment by sharwell** [comment]
Based on the size of this added library (26KiB), I think we have reached a point where the wins may justify the inclusion.

/cc @jinujoseph @jasonmalinowski @jaredpar

**Comment by Pilchie** [comment]
We need to be careful to test this on all of .NET 4.6.1/4.6.2/4.7/4.7.1/4.7.2 and make sure that it behaves as expected.  I know CPS just removed a use of ValueTuple due to binding redirect problems.

**Comment by jaredpar** [comment]
> I know CPS just removed a use of ValueTuple due to binding redirect problems.

Can you elaborate on this? Haven't heard of this before and it's not one of the problems I'm aware of with binding redirects.

**Comment by tmat** [comment]
If we are going to add a new dependency to the compilers and go thru the exercise of updating all the setup authoring and RPS then I suggest we also add a reference to Microsoft.DiaSymReader and delete the copy of the PDB writer from the compiler. See https://github.com/dotnet/roslyn/pull/22611.

**Comment by jaredpar** [comment]
@tmat the other component to updating dependencies is getting permission from DevDiv. In this case @sharwell is trying to trade off memory savings for a new image reference. Need to clear that permission hurdle for Microsoft.DiaSymReader as well.

**Comment by tmat** [comment]
@jaredpar I understand. Just want to use this opportunity to reduce setup authoring work, if possible.

**Comment by Pilchie** [comment]
I sent some email offline.

File: src/Workspaces/Core/Portable/Workspace/Solution/Document.cs [hunk 2]
```diff
@@ -179,7 +179,7 @@ public Task<SyntaxTree> GetSyntaxTreeAsync(CancellationToken cancellationToken =
             }
 
             // do it async for real.
-            return DocumentState.GetSyntaxTreeAsync(cancellationToken);
+            return DocumentState.GetSyntaxTreeAsync(cancellationToken).AsTask();
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/Document.cs]: You should probably update src/NuGet/Microsoft.CodeAnalysis.Workspaces.Common.nuspec to state the dependency too. It's "fine" for now in that workspaces depends on common, and it depends on extensions, but if ever somebody "fixes" the core one to not depend on it, this is now broken.

**Comment by sharwell** [comment]
@jasonmalinowski The problems appeared to involve ValueTuple, not ValueTask. We are already using ValueTuple without problems so whatever happened with the other project doesn't seem to be impacting us.

**Comment by sharwell** [comment]
@jaredpar Any remaining issues from your side?

File: src/Compilers/Core/Portable/CodeAnalysis.csproj [hunk 3]
```diff
@@ -45,6 +45,7 @@
     <PackageReference Include="System.Collections.Immutable" Version="$(SystemCollectionsImmutableVersion)" />
     <PackageReference Include="System.Reflection.Metadata" Version="$(SystemReflectionMetadataVersion)" />
     <PackageReference Include="System.Text.Encoding.CodePages" Version="$(SystemTextEncodingCodePagesVersion)" />
+    <PackageReference Include="System.Threading.Tasks.Extensions" Version="$(SystemThreadingTasksExtensionsVersion)" />
```
Comment (khellang) [inline src/Compilers/Core/Portable/CodeAnalysis.csproj]: You could update the `PerformanceSensitiveAttribute`'s docs to include a proper `<see cref="ValueTask" />` tag now 😉 

https://github.com/dotnet/roslyn/blob/682208901aee33a1d322bc5fdf488106e3a3cfb9/src/Compilers/Core/Portable/InternalUtilities/PerformanceSensitiveAttribute.cs#L82

**Comment by jaredpar** [comment]
@sharwell need to add the version string for your new package here

https://github.com/dotnet/roslyn/blob/master/src/NuGet/NuGetProjectPackUtil.csproj#L93

The error message is definitely not helpful. I ran into it frequently when working with `dotnet pack`. I've filed the following issue on NuGet to make this clearer. For instance at least tell me the line number where the version string was wrong. 

https://github.com/NuGet/Home/issues/6806

**Comment by sharwell** [comment]
@Pilchie This change trades a one-time 28KiB assembly load for the savings described at the top, and allows us to make similar changes moving forward.

**Comment by Pilchie** [comment]
I think this is probably worthwhile, but it will likely still need an exception.  Let's wait for Monday, and loop in @MattGertz about the right time to take it (since merging now might block insertions of other changes to `master`).

**Comment by sharwell** [comment]
@dotnet-bot retest this please

**Pull request merged**